pkg/component/runtime/manager_test.go

```diff
@@ -3481,6 +3481,164 @@ LOOP:
+	require.NoError(t, err)
+	errCh := make(chan error)
+	go func() {
+		err := m.Run(ctx)
+		if errors.Is(err, context.Canceled) {
+			err = nil
+		}
+		errCh <- err
+	}()
+
+	waitCtx, waitCancel := context.WithTimeout(ctx, 1*time.Second)
+	defer waitCancel()
+	if err := waitForReady(waitCtx, m); err != nil {
+		require.NoError(t, err)
+	}
+
+	binaryPath := testBinary(t, "component")
+	comp := component.Component{
+		ID: "fake-default",
+		InputSpec: &component.InputRuntimeSpec{
+			InputType:  "fake",
+			BinaryName: "",
+			BinaryPath: binaryPath,
+			Spec:       fakeInputSpec,
+		},
+		Units: []component.Unit{
+			{
+				ID:       "fake-input-1",
+				Type:     client.UnitTypeInput,
+				LogLevel: client.UnitLogLevelTrace,
+				Config: component.MustExpectedConfig(map[string]interface{}{
+					"type":    "fake",
+					"state":   int(client.UnitStateHealthy),
+					"message": "Fake Healthy 1",
+				}),
+			},
+			{
+				ID:       "fake-input-2",
+				Type:     client.UnitTypeInput,
+				LogLevel: client.UnitLogLevelTrace,
+				Config: component.MustExpectedConfig(map[string]interface{}{
+					"type":    "fake",
+					"state":   int(client.UnitStateHealthy),
+					"message": "Fake Healthy 2",
+				}),
+			},
+			{
+				ID:       "fake-input-3",
+				Type:     client.UnitTypeInput,
+				LogLevel: client.UnitLogLevelTrace,
+				Config: component.MustExpectedConfig(map[string]interface{}{
+					"type":    "fake",
+					"state":   int(client.UnitStateHealthy),
+					"message": "Fake Healthy 3",
+				}),
+			},
+			{
+				ID:       "fake-input-4",
+				Type:     client.UnitTypeInput,
+				LogLevel: client.UnitLogLevelTrace,
+				Config: component.MustExpectedConfig(map[string]interface{}{
+					"type":    "fake",
+					"state":   int(client.UnitStateHealthy),
+					"message": "Fake Healthy 4",
+				}),
+			},
+		},
+	}
+
+	subCtx, subCancel := context.WithCancel(context.Background())
+	defer subCancel()
+	subErrCh := make(chan error)
+	go func() {
+		sub := m.Subscribe(subCtx, "fake-default")
+		for {
+			select {
+			case <-subCtx.Done():
+				return
+			case state := <-sub.Ch():
+				t.Logf("component state changed: %+v", state)
+				if state.State == client.UnitStateFailed {
+					subErrCh <- fmt.Errorf("component failed: %s", state.Message)
+				} else {
+					healthyCount := 0
+					stoppedCount := 0
+					for _, unit := range state.Units {
+						if unit.State == client.UnitStateFailed {
+							subErrCh <- fmt.Errorf("unit failed: %s", unit.Message)
+						} else if unit.State == client.UnitStateHealthy {
+							healthyCount += 1
+						} else if unit.State == client.UnitStateStopped {
+							stoppedCount += 1
+						} else if unit.State == client.UnitStateStarting {
+							// acceptable
+						} else {
+							// unknown state that should not have occurred
+							subErrCh <- fmt.Errorf("unit reported unexpected state: %v", unit.State)
+						}
+					}
+					if healthyCount == 4 {
```

Comment:
Is there a way that you can assert that these units were received via the chunking protocol?

If there were a bug that caused the grpcConfig above not to apply, then this test would still pass and we wouldn't know if it were testing the chunking protocol or not.

Reply:
I understand the reasoning for wanting to confirm this, but I believe we would really be adding lots of extras in the code path to test that. Its rather deep into the communication had basically handled all transparently by the elastic-agent-client library.

I think to really have something to verify that it would require modification to the elastic-agent-client and they would have to be public hooks, or callbacks to track that it was sent/receive chunked.

Reply:
I get that it's more work, but I don't like that this test could still pass without the feature actually doing anything.

If we added atomic stats counters to the client, and added some for the number of chunked messages sent/received, then you'd probably be able to confirm it was working pretty easily.

This would generally be useful to expose via the agent diagnostics and HTTP interface one day as well.

Reply:
Thats exploding this bug fix into adding metrics to the elastic-agent-client, seems very unrelated and unreasonable.

Another more isolated option is make the check-in messages always larger than the default GRPC configuration message size, and remove the setting of that option in the test. That way if that value is changed the test will always create a message size that is larger than that amount to ensure that it always still works. The test can also ensure the minimum size is 4MB which is GRPC default, not Agent default to ensure that if the default is ever broken it still works.

Reply:
That works, it's the least disruptive to the rest of the PR.

I still think it would be nice to have a way to tell that chunking is happening or has happened so we can isolate possible bugs to it, but I agree that is more work.

Reply:
Added to the PR. It found an issue with the chunking function, as it was not taking into the account of overhead for each element in an repeated structure in. https://github.com/elastic/elastic-agent-client/issues/94 will need to be fixed before this PR will pass the unit test.

Reply:
The fix in elastic-agent-client has been released and I have updated the PR to use it.